### PR TITLE
many: finish `snap set` API.

### PR DESCRIPTION
**Pull request by kyrofa**
This PR makes more progress on LP: [#1596629](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/1596629) by adding support for an initial context for the hook (and handler), and making use of it in the `snap set` API.

### CODE REVIEW

File: overlord/configstate/applyconfighandler.go [hunk 1]
```diff
@@ -19,7 +19,15 @@
...
-import "github.com/snapcore/snapd/overlord/hookstate"
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/overlord/hookstate"
+)
+
+// CachedTransaction is the index into the context cache where the initialized
+// transaction is stored.
+type CachedTransaction struct{}
```
Comment (niemeyer) [inline overlord/configstate/applyconfighandler.go]: This is probably a better building block, and voids exposing the empty struct trick as a public API:
```
func ContextTransaction(context *hookstate.Context) *Transaction {
        t, ok := context.Cached(cachedTransaction{}).(*Transaction)
        if ok {
                return t
        }
        t := NewTransaction(context.State())
        context.Cache(cachedTransaction{}, t)
        return t
}
```
Note it already relies on the changes proposed in #2014.

File: overlord/configstate/tasksets.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/overlord/hookstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+)
+
+// ApplyConf returns a taskset required to apply the given configuration
+// patch.
+func ApplyConf(s *state.State, snapName string, patchValues map[string]interface{}) *state.TaskSet {
```
Comment (niemeyer) [inline overlord/configstate/tasksets.go]: `Change` might be a better name here, becoming `configstate.Change`, and creating the `change-config` task, firing the `config-changing` hook.

File: overlord/hookstate/hookmgr.go [hunk 3]
```diff
@@ -85,10 +85,16 @@ func Manager(s *state.State) (*HookManager, error) {
...
+// initial context must properly marshal and unmarshal with encoding/json.
+func HookTask(s *state.State, taskSummary, snapName string, revision snap.Revision, hookName string, initialContext map[string]interface{}) *state.Task {
 	task := s.NewTask("run-hook", taskSummary)
-	task.Set("hook-setup", HookSetup{Snap: snapName, Revision: revision, Hook: hookName})
+	task.Set("hook-setup", HookSetup{
+		Snap:     snapName,
+		Revision: revision,
+		Hook:     hookName,
+	})
+	task.Set("initial-context", initialContext)
```
Comment (niemeyer) [inline overlord/hookstate/hookmgr.go]: Why isn't this setting "hook-context" directly?

File: overlord/hookstate/hookmgr.go [hunk 4]
```diff
@@ -139,11 +145,17 @@ func (m *HookManager) Context(contextID string) (*Context, error) {
 func (m *HookManager) doRunHook(task *state.Task, tomb *tomb.Tomb) error {
 	task.State().Lock()
 	setup := &HookSetup{}
-	err := task.Get("hook-setup", setup)
+	hookSetupError := task.Get("hook-setup", setup)
+	var initialContext map[string]interface{}
+	initialContextError := task.Get("initial-context", &initialContext)
```
Comment (niemeyer) [inline overlord/hookstate/hookmgr.go]: None of the changes here are necessary if we just use hook-context directly.

File: tests/lib/snaps/snapctl-hooks/meta/hooks/apply-config [hunk 5]
```diff
@@ -1,37 +1,60 @@
...
-	exit 1
-fi
+	expected_output='{\n"foo": "bar"\n}'
+	if "$output" -ne "$expected_output"; then
+		echo "Expected output to be '$expected_output', but it was '$output'"
+		exit 1
+	fi
+}
+
+command=$(snapctl get command)
```
Comment (niemeyer) [inline tests/lib/snaps/snapctl-hooks/meta/hooks/apply-config]: \o/

File: overlord/configstate/applyconfighandler.go (hunk 1 above)
Comment (kyrofa) [inline overlord/configstate/applyconfighandler.go]: Yeah good call, okay, in the latest push I essentially moved `getTransaction` to `ContextTransaction`.

File: overlord/configstate/tasksets.go (hunk 2 above)
Comment (kyrofa) [inline overlord/configstate/tasksets.go]: Sure, done.

File: overlord/hookstate/hookmgr.go (hunk 3 above)
Comment (kyrofa) [inline overlord/hookstate/hookmgr.go]: The encapsulation issues we discussed on telegram, but I've made this change.

File: overlord/hookstate/hookmgr.go (hunk 4 above)
Comment (kyrofa) [inline overlord/hookstate/hookmgr.go]: Done.

File: overlord/configstate/tasksets_test.go [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+	s.state = state.New(nil)
+}
+
+func (s *tasksetsSuite) TestChange(c *C) {
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	taskset := configstate.Change(s.state, "test-snap", map[string]interface{}{
+		"foo": "bar",
+	})
```
Comment (pedronis) [inline overlord/configstate/tasksets_test.go]: doesn't seem to test the initialContext

File: overlord/configstate/applyconfighandler.go [hunk 7]
```diff
@@ -26,12 +26,54 @@ type applyConfigHandler struct {
...
 
+// cachedTransaction is the index into the context cache where the initialized
+// transaction is stored.
+type cachedTransaction struct{}
+
+// ContextTransaction retrieves the transaction cached within the context (and
+// creates one if it hasn't already been cached).
+func ContextTransaction(context *hookstate.Context) *Transaction {
+	context.Lock()
+	defer context.Unlock()
```
Comment (pedronis) [inline overlord/configstate/applyconfighandler.go]: as usual I would leave the locking to the callers

File: overlord/configstate/tasksets_test.go (hunk 6 above)
Comment (kyrofa) [inline overlord/configstate/tasksets_test.go]: Indeed, good catch. Done.

File: overlord/configstate/applyconfighandler.go (hunk 7 above)
Comment (kyrofa) [inline overlord/configstate/applyconfighandler.go]: Very well, done.

**Pull request merged**